### PR TITLE
Add sections on the Web being safe to browse, trusted user interface, and asking users for consent.

### LINKED ISSUE #146: Don't Erode Users' Trust in the Web
@dbaron wrote, in [a comment](https://github.com/w3ctag/security-questionnaire/issues/48#issuecomment-495566652) on w3ctag/security-questionnaire#48:

> I also think there's a third piece that needs to be considered:  not just the specific security risks added from one specific feature, but the effect of adding those security risks to the overall picture that users have about the idea that it is safe to visit websites.  A big piece of the web's value is that it is safe for users to visit websites.  In order for this value to exist, users must *understand* that it is safer to visit websites than to install apps or software, and act based on that understanding.  Doing things that complicate that understanding or complicate what users need to understand about the safety of the web (e.g., adding features that are less safe) reduces the ability of users to act based on that understanding of safety, or to act in ways that correctly reflect the safety that exists.

[I think](https://twitter.com/hober/status/1134509819937906688) this is a sufficiently important insight that we should capture it somewhere, either here in our Design Principles (this issue), in our Ethical Web Principles finding (w3ctag/ethical-web-principles#20), or in a resurrected HTML Design Principles finding (w3ctag/design-reviews#426).

### PR DESCRIPTION
**Pull request by dbaron**
This is a very rough draft of changes to add sections on the Web being safe to browse, trusted user interface, and asking users for consent.  It is intended to fix #146, though it's possible I've gone a little bit beyond my remit in that issue.

This contains a bunch of ideas that I think are widely accepted, but also some that are probably less so, and some that could certainly use more research on what users *actually* understand about the web's safety.  (I thought about whether to explicitly call out the desire for research on that, but so far haven't done so.)

I think this will likely need a considerable amount of review and refinement.  I suspect if I kept it on my computer until next week and then looked at it again, I'd want to rewrite most or all of it.  But nevertheless I think it's useful to get some feedback on it, particularly on the high-level aspects:  are these the right topics to bring up here, the right way to organize them, and the right points to make?

I've picked a few reviewers who I think are likely to have comments, but I'd welcome comments from others as well.

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+
+<h3 id="safe-to-browse">It should be safe to visit a web page</h3>
+
+Hyperlinks, links from one page to another, are one of the foundations of the Web.
+
+Following a link, or visiting a web page, should be safe.
+Users who are following a link to a web page they haven't seen before
+should not have to fear for the security of their computer
+or for essential aspects of their privacy.
+(One caveat here is that users of the Web
```
Comment (alice) [inline index.bs]: I'm not sure I understand this caveat. I agree with the general principle, of course, but I'm not sure what it adds here - to me it reads as though there can/should be escape hatches from users' privacy protections in cases where the user is harming others, which I'm sure is not the intended reading.

File: index.bs [hunk 2]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+should not assume that they are free from consequences
+if their use of the Web is harming others.)
+
+Saying “essential aspects” here skips over quite a bit of detail.
```
Comment (alice) [inline index.bs]: I'm also a little unclear on what this paragraph is saying. What user interfaces are you referring to? Can we rephrase this to focus on the point about user expectations and privacy, and either omit the aspirational aspects or give them their own paragraph?

File: index.bs [hunk 3]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+a person walking down the street generally expects to be recognized by their friends,
+but depending on the country,
+may not expect that they were walking down that street to be recorded in a permanent government database.
+Online, people might have less understanding of what is or isn't possible,
+and their expectations might differ more widely.
+<!-- Can we cite research on this??? -->
+
+When we add new features to the Web
+that might weaken the security or privacy characteristics that the Web currently has,
+we should consider the tradeoffs involved in that particular feature.
```
Comment (alice) [inline index.bs]: Perhaps this is worthy of its own sub-heading?

File: index.bs [hunk 4]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+<h3 id="trusted-ui">Trusted user interface should be trustworthy</h3>
+
+Since users of the Web can follow links to sites they might not know already,
+software that lets them access the Web generally provides user interface
+to show the user who they're interacting with and how,
+such as by displaying part or all of the URL or other information about the site,
+or whether the connection is secure.
+Since users rely on this information to learn what site they're on
+and to make judgments about whether it is trustworthy,
+it is important that sites not be able to spoof or override this user interface.
```
Comment (alice) [inline index.bs]: This seems out of scope for API design, albeit a critical point for user agent UI design.

It seems like the point below is the more critical one: if a feature relies on adding extra decisions for a user to make, care should be taken that the feature is providing enough value (either through providing a clear user benefit, or through simplifying some other decision process for the user) to make it worthwhile.

File: index.bs [hunk 5]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+and be able to consent to doing so
+(even though they might not fully understand the privacy implications of doing so).
+
+We should not depend on asking the user for consent
+(via a permission prompt or other mechanism)
+if we don't have a way to express that request in a way
+that users will understand what is being requested
+and the main implications of giving their consent.
+
+We should also not ask users to consent to something
```
Comment (alice) [inline index.bs]: Is this worth a separate sub-heading as well?

File: index.bs [hunk 6]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+
+<h3 id="informed-consent">Ask users for informed consent when appropriate</h3>
+
+In some cases,
+we might add features to the Web that are not appropriate to allow for any website a user visits,
+but that are reasonable to have if the user gives informed consent.
+An example of such a feature might be support for location:
+many users seem likely to understand what it means to share their current location with a website
+and be able to consent to doing so
+(even though they might not fully understand the privacy implications of doing so).
```
Comment (alice) [inline index.bs]: Is this an editorial on the geolocation API?

File: index.bs [hunk 7]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+
+We should not depend on asking the user for consent
+(via a permission prompt or other mechanism)
+if we don't have a way to express that request in a way
+that users will understand what is being requested
```
Comment (alice) [inline index.bs]: Which users? How do we determine what the "target audience" users will or won't be able to understand?

File: index.bs [hunk 8]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+for something that they can't effectively prevent.
+Doing so would send the wrong message about the safety characteristics of the Web
+and would lead users to be disappointed when the expectations of privacy,
+which they learned from trusted user interface,
+did not match reality.
+
+Asking users for consent via permission prompts
+can both reinforce the idea that the web is safe
+by showing the user that certain things won't happen without their permission,
+and can also show how scary a place the web is
```
Comment (alice) [inline index.bs]: Is this a risk we're warning about?

File: index.bs (hunk 4 above)
Comment (hober) [reply index.bs]: I see your point, @alice, but I think it's in scope to the extent that different API shapes make different UI designs easier / more likely / etc.

Comment (alice) [reply index.bs]: Perhaps that point should be made more explicitly, in that case?

Comment (hober) [reply index.bs]: Agreed!

File: index.bs (hunk 1 above)
Comment (dbaron) [reply index.bs]: So the caveat is really to the sentence before the quote -- the idea that following a link should always be safe.  Taking that position absolutely would imply that we're taking the position that governments shouldn't punish people for, say, viewing child pornography.  And I don't think that's a position that I'd like to take.  So I want some way to say that we want it to be safe from a technical perspective but not necessarily from a legal one.

(At the same time, there's an entirely separate discussion about mechanisms that governments want to use to enforce some legal rules, and what the downsides of those mechanisms are, and while some of that discussion is relevant to the TAG, I'd like to avoid getting into it in this section.)

I should think more about a better way to word this.

File: index.bs (hunk 6 above)
Comment (dbaron) [reply index.bs]: I was trying to say that I think the geolocation API is ok since it's a question that users can understand, even if they don't necessarily follow all the implications of how location tracking is used in today's ecosystem of tracking/sharing/reselling user data.

File: index.bs [hunk 9]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+In doing so, we may want to focus on the cases where
+user expectations of what information is revealed when using the Web
+don't match the technical details.
+We might be able to use the design of user interfaces
+to make it clear what users of the Web should expect,
+since one aspect of privacy problems is when reality doesn't meet expectations.
+For example,
+a person walking down the street generally expects to be recognized by their friends,
+but depending on the country,
+may not expect that they were walking down that street to be recorded in a permanent government database.
```
Comment (hober) [inline index.bs]: this sentence reads strangely to me. I think I know what you're trying to say, but this is a confusing way to say it.

Does this read better to anyone else:

> For example, a person walking down the street generally expects to be recognized by their friends but,
> depending on the country,
> they may not expect the fact that they walked down that particular street at that particular time to be recorded in a permanent government database.

File: index.bs [hunk 10]
```diff
@@ -44,6 +44,105 @@ url: https://dom.spec.whatwg.org/#ref-for-concept-getelementsbytagname; type: in
...
+and to make judgments about whether it is trustworthy,
+it is important that sites not be able to spoof or override this user interface.
+
+When we add new features to the web,
+we should consider whether they require new complexity in this user interface,
+and whether that complexity would reduce users' ability
+to make correct judgments about who they're interacting with
+and whether the user interface is trustworthy or spoofed.
+
+<h3 id="informed-consent">Ask users for informed consent when appropriate</h3>
```
Comment (hober) [inline index.bs]: Internally on the WebKit team, we tend to talk about "meaningfully informed consent", by which we mean that not only was the user informed of the risks, they were informed of the risks in a way we believe they're likely to be able to understand and make sound judgments based upon. I don't know if this is a distinction worth making here.

File: index.bs (hunk 5 above)
Comment (hober) [reply index.bs]: Does this paragraph argue against the existence of private browsing mode? Users sometimes think that there will be no *server-side* record of what they did while visiting a site with private browsing mode on.

File: index.bs (hunk 4 above)
Comment (alice) [reply index.bs]: https://github.com/w3ctag/design-reviews/issues/461#issuecomment-587618596 this seems like a good first pass.

File: index.bs (hunk 3 above)
Comment (dbaron) [reply index.bs]: Sorry for the delay getting back to these -- I'm going to post revisions soon (I hope) that will address many of the comments.

But while I'm working on that, I wanted to get back to this one:  I think this paragraph is still pretty central to the idea that it should be safe to visit a web page, and I think I'd like to keep them together in the same subsection.  (But perhaps some of the other edits I'm currently making will make this a little clearer.)

File: index.bs (hunk 4 above)
Comment (dbaron) [reply index.bs]: I'm a little confused by @alice's last comment in this thread -- was it meant as a comment on the following section (on informed consent), rather than this section on trusted UI?

(Also, I think the "point below" in the initial comment was referring to the second paragraph of this section, and not the following section, but please correct me if I'm wrong.)

In any case, what I've tried to do here (and I realize I should comment before pushing the changes since pushing the changes will likely hide the review comments) in convert these two paragraphs into one and connect them more closely.  Though perhaps you were looking for something different?

File: index.bs (hunk 5 above)
Comment (dbaron) [reply index.bs]: To @alice's question of separate subheading starting with the third paragraph of this section -- I think both the second and third paragraphs of this section are trying to define the "when appropriate" in the section's title, so I prefer not.

Regarding @hober's question on private browsing mode:  I don't think it argues against the existence of such a mode, but I would argue (although this is more tied to the first of the three subsections) that implementations offering such a mode ought to try to adjust its user interface so that users have more accurate expectations of what it does and doesn't do.

File: index.bs (hunk 7 above)
Comment (dbaron) [reply index.bs]: I'll expand on this based on https://github.com/w3ctag/design-reviews/issues/461#issuecomment-587852956 and https://github.com/w3ctag/design-reviews/issues/461#issuecomment-587895409 .

File: index.bs (hunk 8 above)
Comment (dbaron) [reply index.bs]: Maybe.  I fully admit it's a little bit off-topic for the section, but I think it's a risk worth mentioning.  But I'm open to taking it out if you think it doesn't fit.

File: index.bs (hunk 2 above)
Comment (dbaron) [reply index.bs]: I've tried to do this; I think I pulled the aspirational bits out into their own paragraph, which I actually ended up structuring as a list.

File: index.bs (hunk 9 above)
Comment (dbaron) [reply index.bs]: I agree it's better.  I took most of your suggestions but turned "depending on the country" into a parenthetical and omitted a few words ("the fact", "particular", "particular").

File: index.bs (hunk 10 above)
Comment (dbaron) [reply index.bs]: Maybe, though I'm actually worried, based on its definition in other contexts (particularly medical ones), that the term "informed consent" might even be too strong, since it seems to imply being properly informed of all risks.  Does "meaningful consent" make sense?  I've left it alone for now, but I'm definitely open to changing it.

**Comment by dbaron** [comment]
Also, I've made [a live version of the changes on this branch](https://dbaron.github.io/design-principles/#api-principles) visible.

File: index.bs (hunk 10 above)
Comment (alice) [reply index.bs]: I like "meaningful consent"

File: index.bs (hunk 6 above)
Comment (alice) [reply index.bs]: I think the parenthetical might be best as its own clarifying sentence, in that case.

File: index.bs (hunk 8 above)
Comment (alice) [reply index.bs]: Perhaps it should be rewritten as two sentences illustrating two sides of a trade-off.

File: index.bs (hunk 6 above)
Comment (dbaron) [reply index.bs]: Perhaps, but I'm going to leave this one as is for now.

File: index.bs (hunk 8 above)
Comment (dbaron) [reply index.bs]: I changed it to 2 sentences.

File: index.bs (hunk 10 above)
Comment (dbaron) [reply index.bs]: I changed it to "meaningful consent" (and the `id` to "consent").
